### PR TITLE
NNF Element Controller Recovery

### PR DESCRIPTION
**Pull request by NateRoiger**
Introduces a new CRD, NnfNodeECData (sorry Tony) for holding the critical data for nnf-ec. The corresponding controller implements the StorageProvider interface.

Had to reshuffle where nnf-ec starts to ensure that the StorageProvider (the reconciler) is assigned before nnf-ec.Init(). This caused a couple changes to the NnfNode and NnfNodeStorage so they delay in accessing the NNF Storage Service until it is ready. 

I also retired the service/endpoint from NnfNode code - nobody uses the service name anyways and it just complicated things.

Signed-off-by: Nate Roiger <nate.roiger@hpe.com>

### CODE REVIEW

**Review by ajfloeder — APPROVED** [review]

**Review by ajfloeder — APPROVED** [review]

File: controllers/nnf_node_ec_data_controller.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			}
+
+			if err := r.Create(ctx, &data); err != nil {
+				return err
+			}
+		}
+	}
+
+	// This resource acts as the storage provider for the NNF Element Controller
+	persistent.StorageProvider = r
```
Comment (matthew-richerson) [inline controllers/nnf_node_ec_data_controller.go:82]: Do you need the whole reconciler here, or just the client?

File: controllers/nnf_node_ec_data_controller.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
+/*
```
Comment (matthew-richerson) [inline controllers/nnf_node_ec_data_controller.go:1]: It seems weird to me that this code is in nnf-sos and is part of the controller framework. It's not a reconciler, and it's not part of nnf-sos. Is there something preventing you from making a k8s client down in nnf-ec? Or passing the existing k8s client from nnf-sos to nnf-ec if you wanted to re-use it?

**Review by matthew-richerson — APPROVED** [review]

File: controllers/nnf_node_ec_data_controller.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+		return err
+	}
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NnfNodeECDataReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.Add(r); err != nil {
```
Comment (roehrich-hpe) [inline controllers/nnf_node_ec_data_controller.go:198]: Add a comment like this one that you have in nnf_node_controller:
```// Causes NnfNodeReconciler.Start() to be called.```
Sometimes these clues help to stitch things together;  in this case it takes a bit to work it out even with the help of VSCode.

File: controllers/nnf_node_ec_data_controller.go (hunk 2 above)
Comment (NateRoiger) [reply controllers/nnf_node_ec_data_controller.go:1]: I was trying to preserve nnf-ec's lack of dependency on kubernetes. That way nnf-ec can continue to be an independent program.

I suppose I could pass in the k8sClient/NamespacedName and in nnf-ec implement the CRD Storage Provider... I'm not sure where the CRD would live then; and how would it get deployed?

File: controllers/nnf_node_ec_data_controller.go (hunk 1 above)
Comment (NateRoiger) [reply controllers/nnf_node_ec_data_controller.go:82]: client & namespaced name

File: controllers/nnf_node_ec_data_controller.go (hunk 2 above)
Comment (roehrich-hpe) [reply controllers/nnf_node_ec_data_controller.go:1]: A CRD doesn't need to be paired with a reconciler.  There is  an option on kubebuilder/operator-sdk  to say you won't want the reconciler.

Comment (NateRoiger) [reply controllers/nnf_node_ec_data_controller.go:1]: I do think we'll eventually want code in reconciler strictly for delete handling in place of `nnf-ec --purge`. Using purge only deletes namespaces and is therefore problematic.

**Review by roehrich-hpe — APPROVED** [review]

**Pull request merged**